File: 1448. Count_Good_Nodes_in_Binary_Tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    int DFS(TreeNode* cur_node, int cur_num){
        
        if(cur_node == nullptr)
            return 0;
        
        int tmp_res = 0;
        
        int next_num = cur_num;
        if(cur_node->val >= cur_num){
            next_num = cur_node->val;
            tmp_res++;
        }
        
        return tmp_res + DFS(cur_node->left, next_num) + DFS(cur_node->right, next_num);
        
    }
    
    int goodNodes(TreeNode* root) {
        
        return DFS(root, root->val);
        
    }
};

File: 1448. Count_Good_Nodes_in_Binary_Tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    int DFS(TreeNode cur_node, int cur_num){
        
        int tmp_res = 0;
        
        int next_num = cur_num;
        if(cur_node.val >= cur_num){
            next_num = cur_node.val;
            tmp_res++;
        }
        
        if(!Objects.isNull(cur_node.left))
            tmp_res += DFS(cur_node.left, next_num);
        if(!Objects.isNull(cur_node.right))
            tmp_res += DFS(cur_node.right, next_num);
        
        return tmp_res;
        
    }
    
    
    public int goodNodes(TreeNode root) {
        return DFS(root, root.val);
    }
}